Kernel Provisioning - initial implementation (#612)
jupyter/jupyter_client: docs/provisioning.rst

be responsible for launching a kernel within its own Kubernetes pod,
communicating the kernel's connection information back to the
application (residing in a separate pod), and terminating the pod upon
the kernel's termination. In essence, a kernel provisioner is an
*abstraction layer* between the ``KernelManager`` and today's kernel
*process* (i.e., ``Popen``).

The kernel manager and kernel provisioner relationship
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Prior to this enhancement, the only extension point for customizing a
kernel's behavior could occur by subclassing ``KernelManager``. This
proved to be a limitation because the Jupyter framework allows for a
single ``KernelManager`` class at any time. While applications could
introduce a ``KernelManager`` subclass of their own, that
``KernelManager`` was then tied directly to *that* application and
thereby not usable as a ``KernelManager`` in another application. As a
result, we consider the ``KernelManager`` class to be an
*application-owned entity* upon which application-specific behaviors can
be implemented.

Kernel provisioners, on the other hand, are contained within the

> ``Kernel Manager`` (i.e., a *has-a* relationship) and applications are

Reviewer: Not sure if this is a typo:
```suggestion
``KernelManager`` (i.e., a *has-a* relationship) and applications are
```

agnostic as to what *kind* of provisioner is in use other than what is
conveyed via the kernel's specification (kernelspec). All kernel
interactions still occur via the ``KernelManager`` and ``KernelClient``
classes within ``jupyter_client`` and potentially subclassed by the
application.

> 

Reviewer 2: Does the KernelManager or KernelClient API changes (additional (optional) methods, method signature change...) ? Is there any impact for current consumers of those APIs?

Maybe worth stating this in the documentation?

Author: Nothing changed on `KernelClient`.  As for `KernelManager` there were some changes to "internal" methods and I agree those kinds of changes should be documented (primarily calls to the Popen instance are replaced with calls to the provisioner instance).  As a result, only `KernelManager` subclass authors should be affected (and hopefully minimally).  However, I would suggest those be part of the 7.0 release docs (or whatever major release this falls in) rather than embedded in functional/dev docs pertaining to Kernel Provisioning itself.

edge node of the Slurm cluster. These kinds of requirements can be
mitigated by leveraging applications like `Jupyter Kernel Gateway <https://github.com/jupyter/kernel_gateway>`_ or
`Jupyter Enterprise Gateway <https://github.com/jupyter/enterprise_gateway>`_
where the gateway server resides on the edge
node of (or within) the cluster, etc.

Discovery
~~~~~~~~~

Kernel provisioning does not alter today's kernel discovery mechanism
that utilizes well-known directories of ``kernel.json`` files. Instead,
it optionally extends the current ``metadata`` stanza within the
``kernel.json`` to include the specification of the kernel provisioner
name, along with an optional ``config`` stanza, consisting of
provisioner-specific configuration items. For example, a container-based
provisioner will likely need to specify the image name in this section.
The important point is that the content of this section is
provisioner-specific.

.. code:: JSON

      "metadata": {
        "kernel_provisioner": {
          "provisioner_name": "k8s-provisioner",
          "config": {
              "image_name": "my_docker_org/kernel:2.1.5",
              "max_cpus": 4
          }
        }
      },

> 

Reviewer 2: Would it be possible to have an additional config_schemastanza that would contain a json schema (https://json-schema.org/) of the config?

That schema could be used by consumer application (eg jupyterlab to show a form in a UI).

Author: Yes, I agree, but feel this falls into the _kernel paramertization_ discussion, which should occur once this is merged.  I believe there will need to be two forms of schemas - one schema produced by the kernel itself describing kernel-specific parameters, the other schema produced by the provisioner describing provisioner-specific parameters.

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The impact of Kernel Provisioning is that it enables the ability to
implement custom kernel provisioners to manage a kernel's lifecycle
within any runtime environment. There are currently two approaches by
which that can be accomplished, extending the ``KernelProvisionerBase``
class or extending the built-in class - ``LocalProvisioner``. As more
provisioners are introduced, some may be implemented in an abstract
sense, from which specific implementations can be authored.

Extending ``LocalProvisioner``
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

If you're interested in running kernels locally and yet adjust their
behavior, there's a good chance you can simply extend
``LocalProvisioner`` via subclassing. This amounts to deriving from
``LocalProvisioner`` and overriding appropriate methods to provide your
custom functionality.

In this example, RBACProvisioner will verify whether the current user is
in the role meant for this kernel. If not, an exception will be thrown.

.. code:: python

    class RBACProvisioner(LocalProvisioner):

        role: str = Unicode(config=True)

        async def pre_launch(self, **kwargs: Any) -> Dict[str, Any]:

> 

Reviewer 2: Looks like there is a well-defined lifecycle, cfr `pre_launch`. Is it possible to list those lifecycle methods here (or link them to the API docs, or to the source)?

Author: I'm not sure what you mean by this - nearly every method pertains to the lifecycle of the kernel processing since that is exactly what `KernelManager` (which drives the calls on the provisioner) is responsible for.  Are you suggesting that we document in the provisioner's method where/when they are called from the `KernelManager`?

Could you please provide an example?  I'm happy to add that kind of thing provided it adds value.

I do add the call relationship in the [base provisioner class](https://github.com/jupyter/jupyter_client/pull/612/files#diff-b370bb301ffe7b9de4edb6f77835f50f930c12552fdacd203839f9710daeedb5R144-R161) - which I view as the canonical doc text for all provisioners.

Reviewer 2: > nearly every method pertains to the lifecycle of the kernel processing since that is exactly what KernelManager

Then it is fine to not embed more details in the doc.

> Are you suggesting that we document in the provisioner's method where/when they are called from the KernelManager?

I did some time ago a sequence diagram for the kernel providers (not merged). We can add that kind of details later on if it makes sense.

Author: Yes, that was helpful to the providers' discussion and a provisioner-based version would be helpful as well.

            if not self.user_in_role(self.role):
                raise PermissionError(f"User is not in role {self.role} and "
                                      f"cannot launch this kernel.")

> 

Reviewer 2: Maybe clearly state that the `user_in_role` method has to be implemented by the Provisioner and is not part of the Provisioner API.

Author: That makes sense - update pushed.

            return super().pre_launch(**kwargs)

It is important to note *when* it's necessary to call the superclass in
a given method - since the operations it performs may be critical to the
kernel's management. As a result, you'll likely need to become familiar
with how ``LocalProvisioner`` operates.

Extending ``KernelProvisionerBase``
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

If you'd like to launch your kernel in an environment other than the
local server, then you will need to consider subclassing :class:`KernelProvisionerBase`
directly.  This will allow you to implement the various kernel process
controls relative to your target environment.  For instance, if you
wanted to have your kernel hosted in a Hadoop YARN cluster, you will
need to implement process-control methods like :meth:`poll` and :meth:`wait`
to use the YARN REST API.  Or, similarly, a Kubernetes-based provisioner
would need to implement the process-control methods using the Kubernetes client
API, etc.

> 

Reviewer 2: Are thos `poll`, `wait`... lifecycle methods enforced by the API? Here also, would be good to have a list (or link) to the lifecycle methods (if the number is not too high and does not break the reading flow).

Author: Again, it's not clear to me what you mean - an example will help me.  I also don't understand what you mean by _enforced by the API_.  Sorry.

Reviewer 2: I should look at the code but did not. My question is "Do all remote provisioners need to implement poll, wait...?". Another way to ask is "Are pool, wait... abstract methods that need to be implemented by all remote provisioners?

Author: Ah - yes, abstract methods are used where "enforcement" is required.  These methods are documented on [`KernelProvisionerBase`](https://github.com/kevin-bates/jupyter_client/blob/fc3227ee71ea7db8f54614d5eb58cec99bce98b5/jupyter_client/provisioning/provisioner_base.py#L24) and in the API portion of the provisioner docs.

By modeling the :class:`KernelProvisionerBase` methods after :class:`subprocess.Popen`
a natural mapping between today's kernel lifecycle management takes place.  This,
coupled with the ability to add configuration directly into the ``config:`` stanza
of the ``kernel_provisioner`` metadata, allows for things like endpoint address,
image names, namespaces, hosts lists, etc. to be specified relative to your
kernel provisioner implementation.

The ``kernel_id`` corresponding to the launched kernel and used by the
kernel manager is now available *prior* to the kernel's launch.  This
enables provisioners with a unique *key* they can use to discover and
control their kernel when launched into resource-managed clusters such
as Hadoop YARN or Kubernetes.

.. tip::
    Use ``kernel_id`` as a discovery mechanism from your provisioner!

Here's a prototyped implementation of a couple of the abstract methods
of :class:`KernelProvisionerBase` for use in an Hadoop YARN cluster to
help illustrate a provisioner's implementation.  Note that the built-in
implementation of :class:`LocalProvisioner` can also be used as a reference.

> Notice the internal method `_get_application_id()`.  This method is

Reviewer: ```suggestion
Notice the internal method ``_get_application_id()``.  This method is
```

> what the provisioner uses to determine the YARN application (i.e.,

Reviewer: ```suggestion
what the provisioner uses to determine if the YARN application (i.e.,
```

``jupyter_client.kernel_provisioners``:

::

            'jupyter_client.kernel_provisioners': [
                'rbac-provisioner = acme.rbac.provisioner:RBACProvisioner',
            ],

where:

-  ``rbac-provisioner`` is the *name* of your provisioner and what will
   be referenced within the ``kernel.json`` file
-  ``acme.rbac.provisioner`` identifies the provisioner module name, and
-  ``RBACProvisioner`` is custom provisioner object name
   (implementation) that (directly or indirectly) derives from
   ``KernelProvisionerBase``

Deploying your custom provisioner
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The final step in getting your custom provisioner deployed is to add a
``kernel_provisioner`` stanza to the appropriate ``kernel.json`` files.
This can be accomplished manually or programmatically (in which some
tooling is implemented to create the appropriate ``kernel.json`` file).
In either case, the end result is the same - a ``kernel.json`` file with
the appropriate stanza within ``metadata``. The *vision* is that kernel
provisioner packages will include an application that creates kernel
specifications (i.e., ``kernel.json`` et. al.) pertaining to that
provisioner.

> 

Reviewer 2: Well, I can see a vision where a single `kernel.json` and some `metadata` (the default values), a `config_schema`. The consumer application would be free to use the schema and pass some `user_config` that  would override the default ones. How those `user_config` would be passed is prolly still to be discussed.

In that sense, I wonder if the additional stanza in the `kernel.json`should be:

- `config`-> `default_config`
- `config_schema`

The `user_config` should be not persisted in the kernel.json but could perfectly find its place in e.g. the metadata of the ipynb notebook.

Author: Again, this falls into the parameterization discussion.  However, I believe we'll need to state that any _required properties_ specified in the schema must have a default value since we can't assume that every consumer of a kernelspec is going to be able to provide parameters.

I agree that persisting user-provided values in the notebook itself is reasonable.  We'll likely want some association/namespacing of those attributes to the associated schema as well.  Then, when that notebook is executed with the same provisioner/schema, _those_ values could be pre-filled/used.  Still lots to work out on that front.

Author: Hmm - in thinking about this just now, I think we also need to separate _configuration_ from _parameters_.  With configuration, I view this as more static in nature.  It can be specified in the `kernel.json`, a config file, or something else - but that aspect of things is completely provisioner-specific and the provisioner is responsible for distinguishing configuration from parameters.  

Examples of configuration might be the resource manager hostname for _this_ installation.  In that case, you don't want to expose that to the end-user, but instead, it's a value that is required by every invocation of that particular provisioned kernel.

Reviewer 2: Interested discussion on nomenclature. Let's say

- `resource_manager_hostname` has to be given to the provisioner, but is enforced by the kernelspec author and can not changed by the user
- `memory` has to be given to provisioner, has a default value of `1G` defined by the kernel author, but can be changed by the user with value `1G`, `5G` or `10G` as stated in a json schema defined by the kernel author.

I see them all as the same thing, by some can not be changed by the user.

Thinking loud, I could see name like `system parameter`/ `user parameter`  or `system config` / `user config`.

Alternative to system vs user could be static vs dynamic.

Author: I agree.  All parameters, _system_ or _user_, _static_ or _dynamic_, should be available to everyone.  Where that presentation takes place is, I believe, where the differences lie.  

I view system parameters as static values where you would not want them changed by the user and those would be oriented to the provisioner's underlying configuration system (which may or may not be trailtets-based). In some (typically single-user) cases, _everyone_ can be split into different roles, namely _admin_ and _user_, where a given _user_ is not exposed to a _system_ parameter.  In other cases, you might want various system parameters exposed to the user for modification.  Since any user-exposed parameters would need to be reflected in the schema and since it could be the case that _system parameters_ should also be available to users for modification, perhaps a _meta-propery_ like `is_hidden` is warranted.  This all implies, then, that the schema itself would need to be configurable (which I suppose is already the case in order to set appropriate enumerated types, etc.). 

Note, there's nothing preventing a provisioner from implementing its user-modifiable parameters based on an underlying configuration system (e.g., configurable traits) so long as the _runtime_ values override the configured values.

These are all the kinds of things that will require hashing out during _parameterized kernels_ discussions.

Reviewer 2: > These are all the kinds of things that will require hashing out during parameterized kernels discussions.

Agree with the previous comment. fyi I will start experimenting ui in jlab and push more stanza just to make sense of the parametrisation options. I hope to come with a demo at the next server meeting.
